docs/react/week-3/lesson.md

```diff
@@ -222,7 +360,7 @@ Phew! That was a lot of work just to render an image! But we're not quite done y
         ...
         <button onClick={handle2020Click}>Fetch image for 2020</button>
         ...
-        {date ? <MartianImageFetcher={date} /> : null}
+        {date ? <MartianImageFetcher photoDate={date} /> : null}
```

Comment:
I'm worried that, having not explained any of this (i.e. the dependencies array, or even what a dependency is), bullet point 18 is confusing, and may leave students feeling like this is more complicated than it is just because they're not familiar with the terminology - particularly as the phrasing of the bullet point suggests to me that the reader _should_ already know this, but we didn't explain any of these concepts to them yet.

Perhaps we could add something to the end of the previous "The timeline of this component" para, like "When the response is received, update the state and re-render so that we can show the data on-screen. The way we've called it, `useEffect` _only_ calls its callback function the first time a component renders - when we do the re-render to show the data on-screen, we will _not_ do another `fetch`."

Then we could rephrase this bullet point as:
This time, we do **not** queue an effect. React doesn't know that it should - remember, the point of `useEffect` is to not call the callback function every time the component renders, so we only queued the effect on the **first** render!

Reply:
Thanks, this is good feedback.

I'm a bit in two minds here: I was intentionally avoiding the deps array in the "The timeline of this component" para, because I wanted to get across the "loop" concept of setting state inside the fetch first. I didn't want to get into too much detail that distracts from this point. The "walkthrough" section was designed to introduce the deps array separately.

However, I do agree that it makes point 18 too confusing and out-of-context. I therefore added a sentence below the "The timeline of this component" para which calls out that the effect wasn't run a second time, but says that it'll be covered later.

I also reworded point 18 a bit so that it doesn't explicitly call it the "dependencies array".

Reply:
I think this looks great, a huge improvement!

One thing we may want to do is also add a `console.log("Fetching data from NASA");` line in the `useEffect` callback before the `fetch`, so that we can more obviously demonstrate that the callback isn't being called again. As things currently stand, we say "You might notice that even though we re-rendered, we did not run the useEffect a second time", but we haven't really _shown_ the students that the callback isn't called again (for all they know it is called again, and is returning the same result so not re-rendering, or even _is_ re-rendering just very quickly, or something) - being able to say "The long only happened once, so the function was only called once" could be a nice demonstration.

Another brief thought: I _think_ we've ended up in a place where we don't actually define the word "dependency" any more... We have the text:
```
Because of this behaviour, the second argument to useEffect is called the dependencies argument. We use it whenever we have something in our effect function that depends on a variable outside of the effect function.
```
But I think the first of those sentences pre-supposes being familiar with the concept of a dependency... Maybe we could add a first sentence to the "`useEffect` dependencies array" section, something like: "Our callback function uses the value of the `photoDate` `prop` - what the function does will depend on what value of `photoData` was used, so we call `photoData` a _dependency_ of the calllback function."?

Reply:
> I think we've ended up in a place where we don't actually define the word "dependency" any more...

We have the hidden answer to Exercise D, which has this text:

"We call this a _dependency_ of the effect. Any variables which are used inside the `useEffect` callback function are dependencies."

I'm unsure if this is obvious enough that everyone would find it though?

I originally hid it because it would be trivial to answer Exercise D by just scrolling down a bit. I could just un-hide it and assume that we could have discussion anyway?